dotnet/packages/Microsoft.TeamsAI/Microsoft.TeamsAI/Utilities/JsonConverters/PlanJsonConverter.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System.ComponentModel.Design;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.TeamsAI.AI;
+using Microsoft.TeamsAI.AI.Planner;
+
+namespace Microsoft.TeamsAI.Utilities.JsonConverters
+{
+    internal class PlanJsonConverter : JsonConverter<Plan>
+    {
+        private static JsonEncodedText _typePropertyName = JsonEncodedText.Encode("type");
+        private static JsonEncodedText _commandsPropertyName = JsonEncodedText.Encode("commands");
+
+        public override bool CanConvert(Type objectType)
+        {
+            return typeof(Plan).IsAssignableFrom(objectType);
+        }
+
+        public override Plan? Read(ref Utf8JsonReader reader, Type objectType, JsonSerializerOptions options)
+        {
+
+            if (reader.TokenType != JsonTokenType.StartObject)
```

Comment:
Any need to create `Utf8JsonReader readerClone = reader;` similar to the one in *CommandJsonConverter*?

Just curious, is it because `reader` is a ref and will be passed into child calls so need to create a clone? Correct me if I missed anything.

Reply:
So in this case we don't clone it because in the `JsonSerializer.Deserialize` method call, the reader's pointer is at the start of the command object (within the commands array). 

```
                            v----reader's pointer is here
{ type: 'plan', commands: [ {type: 'DO', action: 'GetItem'} ] }
```

In _CommandJsonConverter_, we want the reader's pointer to be at the start of the object. There's some minor duplication going on there, but I think in this case the code simplicity, trumps the performance hit.

Reply:
Thanks for the explanation. Yep agreed this is a good simple implementation.